### PR TITLE
Allow a callback to be called when we duplicate is found

### PR DESCRIPTION
**Pull request by rcdmrl**

### CODE REVIEW

File: src/main/scala/deduplication/DeduplicationContext.scala [hunk 1]
```diff
@@ -27,6 +27,20 @@ trait DeduplicationContext[F[_], ID, ContextID, A] {
...
+    * If the process has already ran successfully before, it will return the original result of
+    * [[fa]]
+    * Otherwise, it will run [[fa]].
+    *
+    * @param id The id of the process to run.
+    * @param fa The effect to run if the process is new.
+    * @param onDeduplicateDetected An effect that's run when a duplicate is detected. It can be useful for logging.
+    * @return the result of [[fa]]
+    */
+  def protect(id: ID, fa: F[A], onDeduplicateDetected: A => F[Unit]): F[A]
```
Comment (rcdmrl) [inline src/main/scala/deduplication/DeduplicationContext.scala:43]: I played with the idea of just having a single method with a default value, like this:
```scala
def protect(id: ID, fa: F[A], onDeduplicateDetected: A => F[Unit] = _ => Applicative[F].unit): F[A]
```

but the `trait` doesn't require any typeclasses so it doesn't have an instance for `Applicative[F]`.

File: src/it/scala/deduplication/DeduplicationSuite.scala [hunk 2]
```diff
@@ -134,6 +135,23 @@ class DeduplicationSuite extends FunSuite {
     }
   }
 
+  test("should call a callback when a duplicate is found") {
```
Comment (rcdmrl) [inline src/it/scala/deduplication/DeduplicationSuite.scala:138]: Not 100% sure we need an IT test

**Review by d-fernandes — APPROVED** [review]
Looks good to me

**Review by joedarby — APPROVED** [review]
LGTM

**Pull request merged**